java/ql/lib/semmle/code/java/security/SensitiveApi.qll

```diff
@@ -3,17 +3,53 @@
  */
 
 import java
+private import semmle.code.java.dataflow.DataFlow
+private import semmle.code.java.dataflow.ExternalFlow
 
 /**
+ * A node that represents the use of a credential.
+ */
+abstract class CredentialsSinkNode extends DataFlow::Node { }
+
+/**
+ * A node representing a password being passed to a method.
+ */
+class PasswordSink extends CredentialsSinkNode {
+  PasswordSink() { sinkNode(this, "credentials-password") }
+}
+
+/**
+ * A node representing a username being passed to a method.
+ */
+class UsernameSink extends CredentialsSinkNode {
+  UsernameSink() { sinkNode(this, "credentials-username") }
+}
+
+/**
+ * A node representing a cryptographic key being passed to a method.
+ */
+class CryptoKeySink extends CredentialsSinkNode {
+  CryptoKeySink() { sinkNode(this, "credentials-key") }
+}
+
+/**
+ * A node representing a cryptographic initialization vector being passed to a method.
+ */
+class InitializationVectorSink extends CredentialsSinkNode {
+  InitializationVectorSink() { sinkNode(this, "encryption-iv") }
+}
+
+/**
+ * DEPRECATED: Use the `PasswordSink` class instead.
  * Holds if callable `c` from a standard Java API expects a password parameter at index `i`.
  */
-predicate javaApiCallablePasswordParam(Callable c, int i) {
+deprecated predicate javaApiCallablePasswordParam(Callable c, int i) {
```

Comment:
Would it make sense to replace the implementation of these predicates (even if we still deprecate them) to use the new `CredentialsSinkNode` subtypes? That way we'd verify that the new definitions are equivalent to the old ones, and we would be able to directly remove the predicates containing the proto-models-as-data strings, since those are private.

Reply:
I am not opposed to that. The only thing I want to make sure we're ok with there being a slight difference since I removed/reclassified some models that were either irrelevant or better classified differently. But it appears that there has been no difference in results.

And to double check, I'm assuming the new versions of the predicates would look something like this?:
```codeql
exists(DataFlow::Node arg, MethodAccess ma | 
   sinkNode(arg, "credentials-password") | 
   c = ma.getCallee() and arg = ma.getArgument(i)
)
```

Reply:
> The only thing I want to make sure we're ok with there being a slight difference since I removed/reclassified some models that were either irrelevant or better classified differently.

If we want to be safe, we could mention that slight difference in the change notes just as we would do if we had improved the precision of these predicates only (instead of the refactor).

> And to double check, I'm assuming the new versions of the predicates would look something like this?:

I was thinking more of something like:

```ql
/**
 * DEPRECATED: Use the `PasswordSink` class instead.
 * Holds if callable `c` from a standard Java API expects a password parameter at index `i`.
 */
deprecated predicate javaApiCallablePasswordParam(Callable c, int i) {
  exists(PasswordSink sink, MethodAccess ma |
    sink.asExpr() = ma.getArgument(i) and c = ma.getCallee()
  )
}
```